docs/kuneiform/procedures.mdx

```diff
@@ -0,0 +1,336 @@
+about the supported functions, see [Functions](/docs/kuneiform/functions).
+
+```ts
+procedure create_user($name text, $age int) public {
+    $id uuid := generate_uuid_v5('455f60aa-0569-4aaa-8469-63be2ec4dd96'::uuid, @txid)
+    // do something
+    // ...
+    INSERT INTO users (id, name, age)
+    VALUES ($id, $name, $age);
+}
+```
+
+Procedures returning tables can also be joined against:
+
+```ts
+procedure get_users() public view returns table(id uuid, name text, age int) {
+    // ...
+}
+
+procedure join_procedure() public {
+    SELECT * FROM posts AS p
+    INNER JOIN get_users() AS u
+    ON p.author_id = u.id;
+    // ...
+}
+```
+
+### Handling Arrays
+
+Procedures support basic array operations. See [Functions](/docs/kuneiform/functions#array-functions) for more information on array functions.
+
+**Arrays use 1-based indexing.**
+
+```ts
+procedure handle_array() public view returns (val text) {
+    $arr text[] := ['hello', 'world'];
+    $arr := array_append($arr, 'goodbye');
+
+    $len int := array_length($arr); // $len == 3
+
+    // return the first value.
+    return $arr[1];
+}
+```
+
+### If/Then
+
+```ts
+procedure age_conditional($age int) public {
+    if $age < 0 {
+        // do something
+    } elseif $age < 18 {
+        // do something
+    } else {
+        // something else
+    }
+}
+```
+
+### For Loop
+
+For loops can be declared over 3 different targets: arrays, ranges, and SQL statements.
+
+#### Array Loop
+
+```ts
+procedure iterate_over_array($names text[]) public {
+    for $name in $names {
+        // $name is of type `text`
+    }
+}
+```
+
+#### Range Loop
+
+```ts
+procedure iterate_over_range() public {
+    // iterating over all numbers from 1 to 25
+    for $i in 1..25 {
+        // $i is type int
+    }
+}
+
+procedure iterate_range_dynamic($start int, $end int) public {
+    // can also iterate over dynamic ranges
+    for $i in $start..$end {
+        // ...
+    }
+}
+```
+
+#### SQL Loops
+
+You can iterate over any SQL statement as well. The following examples reference an example table `users`, with columns `id`, `name`, and `age`.
+
+```ts
+procedure iterate_over_sql() public {
+    for $row in SELECT id, name, age FROM users {
+        // can access $row.id, $row.name, and $row.age
+    }
+}
+```
+
+Procedures that return tables can also be iterated over:
+
+```ts
+procedure get_users() public view returns table(id uuid, name text, age int) {
+    // ...
+}
+
+procedure iterate_over_users() public {
+    for $row in SELECT * FROM get_users() {
+        // can access $row.id, $row.name, and $row.age
+    }
+}
+```
+
+#### Breaking From a Loop
+
+To conditionally exit a loop, you can use `break`:
+
+```ts
+procedure get_users() public {
+    $count int := 0;
+    for $row in SELECT * FROM users {
+        // do something
+
+        // break if too many records have been read
+        $count := $count + 1;
+        if $count >= 50 {
+            break
```

Comment:
Does this `break` need a semicolon? It feels weird that every other line in this language ends with a semicolon, but `break` does not?